### ISSUE TITLE
Azure Files PV AuthorizationFailure when using advanced networking

### ISSUE THREAD

**Issue by wiebren** [Azure/AKS#804]
Cluster created using these guides
- https://docs.microsoft.com/en-us/azure/aks/aad-integration
- https://docs.microsoft.com/en-us/azure/aks/azure-files-dynamic-pv
- https://docs.microsoft.com/en-us/azure/aks/configure-advanced-networking

Azure Files persistent volume claims fail with the message:
`persistentvolume-controller   (combined from similar events): Failed to provision volume with StorageClass "azurefile": failed to create share kubernetes-dynamic-pvc-xxx in account xxx: failed to create file share, err: storage: service returned error: StatusCode=403, ErrorCode=AuthorizationFailure, ErrorMessage=This request is not authorized to perform this operation.`

The service principal "AzureContainerService" is listed as owner on the storage account, so it should have access.

I first did a setup without advanced networking, in that setup the Azure Files PV without issues.

**Comment by andyzhangx** [comment]
@wiebren it's not related to the service principal, in k8s, it uses storage account name and key to create an azure file in that storage account, could you confirm whether the original storage account key changed or is there any restriction on the storage account?
You could see your error happens here: https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/azure/azure_storage.go#L47

**Comment by andyzhangx** [comment]
@wiebren could you check the activity logs of the storage account? I don't think it's related to using advanced networking.
cc @VybavaRamadoss @RenaShahMSFT do you know how `err: storage: service returned error: StatusCode=403, ErrorCode=AuthorizationFailure, ErrorMessage=This request is not authorized to perform this operation.` error message comes from azure storage?

**Comment by wiebren** [comment]
![screenshot-portal azure com-2019 02 02-10-37-12](https://user-images.githubusercontent.com/1917145/52162578-a1bebe80-26d6-11e9-9971-a13d8154bedd.png)

**Comment by andyzhangx** [comment]
@wiebren and also paste `kubectl describe pvc PVC-NAME`?

**Comment by andyzhangx** [comment]
@wiebren could you also create a file share on that storage account manually?

**Comment by wiebren** [comment]
@andyzhangx Hmm, I now also get "Access denied" when I go to the files section in the storage account. I tough i might have created the account with different settings then before. I just tried StorageV2_ZRS, StorageV2_LRS & Storage_ZRS, all the same issue

```
Name:          production-wiki
Namespace:     wiki
StorageClass:  azurefile
Status:        Pending
Volume:
Labels:        app=production-wiki
               chart=wiki-0.1.0
               heritage=Tiller
               release=production
Annotations:   volume.beta.kubernetes.io/storage-provisioner: kubernetes.io/azure-file
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:
Access Modes:
Events:
  Type     Reason              Age                        From                         Message
  ----     ------              ----                       ----                         -------
  Warning  ProvisioningFailed  3m38s (x15965 over 2d18h)  persistentvolume-controller  (combined from similar events): Failed to provision volume with StorageClass "azurefile": failed to create share kubernetes-dynamic-pvc-xxx in account xxx: failed to create file share, err: storage: service returned error: StatusCode=403, ErrorCode=AuthorizationFailure, ErrorMessage=This request is not authorized to perform this operation.
RequestId:4313d741-b01a-002d-0cdf-ba01ac000000
Time:2019-02-02T10:10:17.4087251Z, RequestInitiated=Sat, 02 Feb 2019 10:10:17 GMT, RequestId=4313d741-b01a-002d-0cdf-ba01ac000000, API Version=, QueryParameterName=, QueryParameterValue=
Mounted By:  production-54cf9f78c-22tpg
```

**Comment by andyzhangx** [comment]
@wiebren pls try with `StorageV1_LRS`, it should work

**Comment by wiebren** [comment]
I am beginning to think the issue might be that the storage account is created with terraform. In manually create accounts I can access the files section, in terraform created accounts I cannot. So I wanted to compare the settings by looking at the generated automation scripts however the exportTemplate call for the terraform created storage account returns with a "500 Internal Server Error"

**Comment by andyzhangx** [comment]
@wiebren it's `Storage (general purpose v1)`, and you don't need to specify a storage account name, k8s will find a suitable storage account that matches skuName in the same resource group. you could use this storage class: https://github.com/andyzhangx/demo/blob/master/pv/storageclass-azurefile.yaml

**Comment by andyzhangx** [comment]
@wiebren `StorageV2 (general purpose v2)` should also work, following combination works:
![image](https://user-images.githubusercontent.com/4178417/52163370-f7698a00-271b-11e9-9f2e-bf9dda40f8ea.png)

**Comment by wiebren** [comment]
@andyzhangx Tried some different things, seems to be network related. I tried setting "Firewalls and virtual networks" of the storage account to allow all access, instead of only my AKS subnet. Now the volume has been created but my storage account is globally accessible. Also a new issue pops up, now mounting the volume fails:

 ```
mount failed: exit status 32
Mounting command: systemd-run
Mounting arguments: --description=Kubernetes transient mount for /var/lib/kubelet/pods/xxx/volumes/kubernetes.io~azure-file/pvc-xxx --scope -- mount -t cifs -o username=xxx,password=xxx,file_mode=0777,dir_mode=0777,vers=3.0 //xxx.file.core.windows.net/kubernetes-dynamic-pvc-xxx /var/lib/kubelet/pods/xxx/volumes/kubernetes.io~azure-file/pvc-xxx
Output: Running scope as unit run-r0b092d9a65314e828042dcc3c611f0dd.scope.
mount error(13): Permission denied
Refer to the mount.cifs(8) manual page (e.g. man mount.cifs)
```

**Comment by andyzhangx** [comment]
@wiebren could you try run `mount -t cifs ...` on one agent VM?

**Comment by wiebren** [comment]
@andyzhangx That helped, it was using a wrong access key, i suspect suff got mixed up by deleting/creating the storage account e few times. I deleted the PV/PVC's and now it mounts.

So any clue on how to get this working with restricted network access to my storage account?

**Comment by andyzhangx** [comment]
I think use `Standard_LRS`  in azure file storage class would work, could you try this azure file class directly: https://github.com/andyzhangx/demo/blob/master/pv/storageclass-azurefile.yaml, it will search a matching storage account for you.

**Comment by andyzhangx** [comment]
And if you want to use a specified storage account, your could create a storage account as below:
![image](https://user-images.githubusercontent.com/4178417/52166436-30b7ef00-2748-11e9-9ad3-0805a2a59024.png)

**Comment by wiebren** [comment]
@andyzhangx Same thing with Standard_LRS, ErrorCode=AuthorizationFailure when I only allow access from my aks subnet and when I allow all traffic it works.

**Comment by andyzhangx** [comment]
@wiebren 
Suddently I know why this `only allow access from selected network` for storage account does not work on AKS, that's because `persistentvolume-controller` is on AKS master node which is not in the selected network, and that's why it could not create file share on that storage account.

And in the near future, I don't think we would support this feature: `only allow access from selected network` for storage account

cc @sriramkomma

**Comment by andyzhangx** [comment]
one workaround is use azure file static provisioning, that is create azure file share by user, and then user provide the storage account and file share in k8s, here is an example: https://docs.microsoft.com/en-us/azure/aks/azure-files-volume
, I think azure file static provisioning would work on this case, while dynamic provisioning won't work

**Comment by andyzhangx** [comment]
By talking with azure file team, currently this create file share action is also considered as `data-path operation`, in the future, this action will **not** be handled as data-path operation. I will let your know when this feature is available.

**Comment by krottiers** [comment]
I'm experiencing exactly the same issue. 

I would find it strange that the master node doesn't have access to the storage account as you can check the box that microsoft services can still access the storage account.

For me this seems like a bug. The functionality to create a dynamic PVC on a vNet integrated is something we really need. Otherwise we would need to script the creation of a file share to be able to assign a PVC when a new application is deployed.

**Comment by hansverbrugge** [comment]
+1
I'm facing the same issue. Whilst the workaround works, or allowing all traffic in the storage account firewall on creation and re-enable when done, this is something we really need, as we automate the creation of the storage account, including setting up access control and the firewall.

**Comment by andyzhangx** [comment]
cc @VybavaRamadoss @RenaShahMSFT

**Comment by cmendible** [comment]
Hi @andyzhangx  I´m using static provisioning and having te same issue. If I `allow all networks` in the storage account the mount works as expected. Once I restrict the traffic, to the storage account, mounting the volumes fails with: `Permission denied`.

I have configured the service endpoints and even have a static public ip (from a working k8s egress service) whitelisted in the storage account's firewall but without luck.

**Comment by andyzhangx** [comment]
> Hi @andyzhangx I´m using static provisioning and having te same issue. If I `allow all networks` in the storage account the mount works as expected. Once I restrict the traffic, to the storage account, mounting the volumes fails with: `Permission denied`.
> 
> I have configured the service endpoints and even have a static public ip (from a working k8s egress service) whitelisted in the storage account's firewall but without luck.

**Comment by cmendible** [comment]
> sudo mount -t cifs

Some how I missed this message. I'll have to try that an come back to you with the results...

**Comment by cmendible** [comment]
@andyzhangx I can confirm it's working with static provisioning. Seems that something was wrong with my SA firewall confi. I also tested with a new cluster and everything is running smooth!

**Comment by andyzhangx** [comment]
@RenaShahMSFT do you have timeline about when create file share action would **not** be considered as data-path operation?

**Comment by RenaShahMSFT** [comment]
> @RenaShahMSFT do you have timeline about when create file share action would not be considered as data-path operation?

We are targeting this in ~Q3 CY 2019

**Comment by Equilibri0** [comment]
any news ?

**Comment by samisq** [comment]
Any update on this issue? We really need this too

**Comment by mturzynski** [comment]
I've just run into this issue. I agree that static provisioning may be a workaround but it's not really convinent.

**Comment by scivm** [comment]
@RenaShahMSFT Initial target was ~Q3 CY 2019.  Do you know when this will get into AKS?

**Comment by scivm** [comment]
@andyzhangx Do you know if we can restrict access to some microsoft network or we have to leave it open to all networks until this is fixed?

**Comment by andyzhangx** [comment]
> @RenaShahMSFT Initial target was ~Q3 CY 2019. Do you know when this will get into AKS?

@scivm
Here is the working on PR: https://github.com/kubernetes/kubernetes/pull/90350, will try to cherry-pick to v1.18 release after this PR merged.
And also you could try https://github.com/kubernetes-sigs/azurefile-csi-driver, since controller components are in AKS node resource group, it won't have this issue now.

**Comment by andyzhangx** [comment]
This issue is already fixed in k8s v1.19.0 and latest azure file CSI driver.

**Comment by andyzhangx** [comment]
close this issue since I have verified that this issue is fixed in AKS 1.19.0 and also with azure file CSI driver
